tests/uncertainty_test.py

```diff
@@ -26,13 +48,145 @@ def test_broadcast_scales_variances_by_new_subspace_volume():
+        broadcast_with_upper_bound_variances(
+            var, template=sc.empty(sizes={'y': 3, 'z': 2})
+        ),
         expected.transpose(['y', 'z', 'x']),
     )
+
+
+def test_broadcast_new_subspace_volume_with_masks():
+    # Data: A A A A A  Template: X
+    #                            X
+    #                            B
+    #                            B
+    #                            B
+    #                            B
+    x = np.linspace(0.0, 1.0, 5)
+    a = sc.array(dims=['x'], values=x, variances=x)
+    b = sc.DataArray(
+        data=sc.ones(sizes={'y': 6}),
+        masks={
+            'm': sc.array(dims=['y'], values=[True, True, False, False, False, False])
+        },
+    )
+    expected = sc.values(b.data) * sc.values(a)
+    expected.variances = (
+        np.broadcast_to(x, expected.shape) * 4
+    )  # 4 non-masked elements in b
+    assert_identical(broadcast_with_upper_bound_variances(a, template=b), expected)
+
+
+def test_broadcast_with_masks_and_a_common_dimension():
+    # Data: A A A A A  Template: B B B B B
+    #                            X X X X X
+    #                            B B B B B
+    #                            B B B B B
+    #                            X X X X X
+    #                            B B B B B
+    x = np.linspace(0.0, 1.0, 5)
+    a = sc.array(dims=['x'], values=x, variances=x)
+    b = sc.DataArray(
+        data=sc.ones(sizes={'y': 6, 'x': 5}),
+        masks={
+            'm': sc.array(dims=['y'], values=[False, True, False, False, True, False])
+        },
+    )
+    expected = sc.values(b.data) * sc.values(a)
+    expected.variances = np.broadcast_to(x, expected.shape) * 4
+    assert_identical(broadcast_with_upper_bound_variances(a, template=b), expected)
+
+
+def test_broadcast_with_masks_and_two_common_dimensions():
+    x = np.linspace(0.0, 5.0, 10).reshape(5, 2)
+    a = sc.array(dims=['x', 'z'], values=x, variances=x)
+    b = sc.DataArray(
+        data=sc.ones(sizes={'y': 6, 'x': 5, 'z': 2}),
+        masks={
+            'm': sc.array(dims=['y'], values=[False, True, True, False, True, False])
+        },
+    )
+    expected = sc.values(b.data) * sc.values(a)
+    expected.variances = np.broadcast_to(x, expected.shape) * 3
+    assert_identical(broadcast_with_upper_bound_variances(a, template=b), expected)
+
+
+def test_broadcast_with_2d_mask():
+    # Data: A A A A A  Template: B X X B B
+    #                            B B B B B
+    #                            X X X X X
+    #                            X X X X X
+    #                            B B B B B
+    #                            B B B B X
+
+    # TODO: It is not so clear what to do here. This would be, for example, a case where
+    # the direct beam depends on wavelength and layer, while the solid angle depends on
+    # layer and straw, and the mask on the solid angle depends on both layer and straw.
+    # Do we just count the number of non-masked elements in the template?
```

Comment:
I think we have to count the combination of masks, i.e., the masks the output would have.

Reply:
So if you had this, you would count only the non-masked pixels in the output?
![IMG_0412](https://github.com/scipp/esssans/assets/39047984/efce0c59-38c1-48ab-8229-371a5c07e7ee)

Reply:
But as I wrote in the comment, what do you do if e.g. the direct beam function depends on `[wavelength, layer]`, and the solid angle depends on `[layer, straw, pixel]`, and the mask on the solid angle also depends on `[layer, straw, pixel]`?

Reply:
Ignore those masks?

Reply:
Let's talk tomorrow, also about the `irreducible_mask`.